x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -420,19 +353,19 @@ export class Authenticator {
   async logout(request: KibanaRequest) {
     assertRequest(request);
 
-    const sessionStorage = this.options.sessionStorageFactory.asScoped(request);
-    const sessionValue = await this.getSessionValue(sessionStorage);
+    const sessionValue = await this.getSessionValue(request);
     if (sessionValue) {
-      sessionStorage.clear();
-
-      return this.providers.get(sessionValue.provider.name)!.logout(request, sessionValue.state);
+      await this.session.clear(request);
+      return this.providers
+        .get(sessionValue.provider.name)!
+        .logout(request, sessionValue.state ?? null);
     }
 
-    const providerName = this.getProviderName(request.query);
-    if (providerName) {
+    const queryStringProviderName = (request.query as Record<string, string>)?.provider;
```

Comment:
__note:__ while testing this functionality I realized that I degraded it bit by bit during a couple of previous releases (shame on me): token and basic providers stopped preserving `next` parameter during logout initiated by the client side `SessionTimeout` service. I restored this functionality in the latest commit and added a base to support this for the rest of the providers as well in the future. It reminded me that we should treat `null` and `undefined` state differently.